src/lib/core/ordered_latch.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include <assert.h>
+#include "fiber.h"
+#include "small/rlist.h"
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/**
+ * Latch of cooperative multitasking environment, which preserves strict order
+ * of fibers waiting for the latch.
+ */
+struct ordered_latch {
+	/** The fiber that locked the latch, or NULL if unlocked. */
+	struct fiber *owner;
+	/** The queue of fibers waiting for the latch. */
+	struct rlist queue;
+};
+
+/**
+ * Initialize the given latch.
+ */
+static inline void
+ordered_latch_create(struct ordered_latch *l)
+{
+	l->owner = NULL;
+	rlist_create(&l->queue);
+}
+
+/**
+ * Destroy the given latch.
+ */
+static inline void
+ordered_latch_destroy(struct ordered_latch *l)
+{
+	assert(l->owner == NULL);
+	assert(rlist_empty(&l->queue));
+	(void)l;
+}
+
+/**
+ * Return true if the latch is locked.
+ */
+static inline bool
+ordered_latch_is_locked(const struct ordered_latch *l)
+{
+	return l->owner != NULL;
+}
+
+/**
+ * Lock the latch.
+ * If it is already locked by another fiber, wait for it.
+ * Locks are acquired in the strict order as they were requested.
+ */
+static inline void
+ordered_latch_lock(struct ordered_latch *l)
+{
+	assert(l->owner != fiber());
+	if (l->owner == NULL && rlist_empty(&l->queue)) {
+		l->owner = fiber();
+		return;
+	}
+
+	struct ordered_latch waiter;
+	waiter.owner = fiber();
+	rlist_add_tail_entry(&l->queue, &waiter, queue);
+	do {
+		fiber_yield();
```

Comment:
Should it be possible to cancel a fiber trying to acquire the latch?
Currently it will simply ignore the cancellation and wait forever potentially.

Reply:
I am not sure… @locker?

Reply:
AFAICS it's impossible to cancel a fiber waiting on a latch in the current implementation, because we make it non-cancellable before yield:
https://github.com/tarantool/tarantool/blob/9c87898ea0fa9d6ffd721187b9c17ddf4d9bc0bc/src/lib/core/latch.h#L131

I don't see why we should allow that in the new implementation.